wordpress-mobile/WordPress-iOS
# Add initial A/B testing support

Part of https://github.com/wordpress-mobile/WordPress-iOS-Shared/issues/278

This PR adds a basic API to retrieve experiments and their variations from the ExPlat.

(TTL, Anonymization, Deanonymization are coming in subsequent PRs)

### To test

There are no visual changes, so you can rely on the tests (or debugging) to make sure it's working and please propose any changes you have in mind. :)

Also, you can just add `ExPlat().refresh()` in `WordPressAppDelegate`.

### Usage

My idea is to use it like:

```swift
switch abTesting.experiment("reader") {
    case .treatment:
        // Show treatment option
    case .control:
        // Show control option
    case .unknown:
        // Experiment doesn't exist or wasn't retrieved
    case .other(let value) where value == "my-value":
        // any other value 
}
```

Overall, usage should be way more simpler, like:

```swift
switch abTesting.experiment("reader") {
    case .treatment:
        // Show treatment option
    default:
        // Show control option
}
```

### Wait, shouldn't that be on WordPress-iOS-Shared/Kit?

Yes. My plan is to have that in a separate library so both WPiOS and WooiOS (and any other app) can make use of it. But we have a few issues:

- Tracks protocol is on Shared
- `WPAnalyticsTrackerAutomatticTracks` on the other hand is in WPiOS (and Woo have its own implementation)
- The logic to request a V2 endpoint is in another library, WordPress-Kit
- I don't want to create a dependency between Shared and Kit neither rely on the implementation of the `WPAnalyticsTrackerAutomatticTracks` here

Given this would require a lot more discussion I decided to just go ahead and implement it in WPiOS, then we can manage the better way to extract it. (@astralbodies I could really use your expertise here)

PR submission checklist:
